### PR TITLE
Implement fused_q_indexer_rope_hadamard_quant with SYCL

### PR DESCRIPTION
**Pull request by cyxlily**
## Motivation

Add a SYCL implementation of `fused_q_indexer_rope_hadamard_quant` to replace the PyTorch fallback on XPU.
This fused operator combines RoPE, Hadamard transform, and quantization (with weights_out scaling) in a single kernel pass. Based on the CUDA implementation: https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/jit/csrc/deepseek_v4/main_norm_rope.cuh#L458

## Modifications

- Implemented SYCL kernel path for fused_q_indexer_rope_hadamard_quant.
- Registered the op `fused_q_indexer_rope_hadamard_quant` and expose it through sgl_kernel. 
- Added unit tests: tests/test_fused_q_indexer_rope_hadamard_quant.py
- Added benchmark: benchmark/bench_fused_q_indexer_rope_hadamard_quant.py

## Accuracy Tests

Tested fused_q_indexer_rope_hadamard_quant (SYCL) vs PyTorch reference on Intel XPU.
All unit tests passed.

Command: pytest tests/test_fused_q_indexer_rope_hadamard_quant.py
Results: 
========================================================================== test session starts ==========================================================================
platform linux -- Python 3.12.0, pytest-9.1.1, pluggy-1.6.0
rootdir: /home/lily/sgl-kernel-xpu
configfile: pyproject.toml
plugins: anyio-4.14.0
collected 3 items

tests/test_fused_q_indexer_rope_hadamard_quant.py ...                                                                                                             [100%]

=========================================================================== 3 passed in 9.72s ===========================================================================

## Performance

SYCL is faster than the Torch reference across all reported configurations.

fused_q_indexer_rope_hadamard_quant: sycl vs torch
|   batch_size |   num_heads |   max_pos |   sycl_ms |   torch_ms |   speedup |
|-------------:|------------:|----------:|----------------:|-----------:|----------:|
|            1 |           4 |       256 |        0.006032 |   0.078026 |    12.935 |
|            1 |           4 |       512 |        0.005902 |   0.077766 |    13.176 |
|            1 |           4 |      1024 |        0.005772 |   0.077688 |    13.459 |
|            1 |           4 |      2048 |        0.006032 |   0.077688 |    12.879 |
|            1 |           8 |       256 |        0.005824 |   0.081068 |    13.92  |
|            1 |           8 |       512 |        0.005928 |   0.081328 |    13.719 |
|            1 |           8 |      1024 |        0.005824 |   0.081016 |    13.911 |
|            1 |           8 |      2048 |        0.00598  |   0.081276 |    13.591 |
|            1 |          16 |       256 |        0.005824 |   0.09022  |    15.491 |
|            1 |          16 |       512 |        0.005824 |   0.090194 |    15.487 |
|            1 |          16 |      1024 |        0.005876 |   0.090064 |    15.327 |
|            1 |          16 |      2048 |        0.005824 |   0.090272 |    15.5   |
|            1 |          32 |       256 |        0.006136 |   0.095706 |    15.597 |
|            1 |          32 |       512 |        0.006136 |   0.095732 |    15.602 |
|            1 |          32 |      1024 |        0.006136 |   0.095732 |    15.602 |
|            1 |          32 |      2048 |        0.006448 |   0.095992 |    14.887 |
|            4 |           4 |       256 |        0.005928 |   0.091    |    15.351 |
|            4 |           4 |       512 |        0.005876 |   0.090688 |    15.434 |
|            4 |           4 |      1024 |        0.00598  |   0.090896 |    15.2   |
|            4 |           4 |      2048 |        0.005876 |   0.090688 |    15.434 |
|            4 |           8 |       256 |        0.006136 |   0.095472 |    15.559 |
|            4 |           8 |       512 |        0.00624  |   0.095524 |    15.308 |
|            4 |           8 |      1024 |        0.006188 |   0.095524 |    15.437 |
|            4 |           8 |      2048 |        0.006344 |   0.095576 |    15.066 |
|            4 |          16 |       256 |        0.006968 |   0.1001   |    14.366 |
|            4 |          16 |       512 |        0.007124 |   0.100152 |    14.058 |
|            4 |          16 |      1024 |        0.00702  |   0.099996 |    14.244 |
|            4 |          16 |      2048 |        0.007072 |   0.10036  |    14.191 |
|            4 |          32 |       256 |        0.007072 |   0.106964 |    15.125 |
|            4 |          32 |       512 |        0.007072 |   0.107146 |    15.151 |
|            4 |          32 |      1024 |        0.007072 |   0.10712  |    15.147 |
|            4 |          32 |      2048 |        0.006968 |   0.106808 |    15.328 |
|            8 |           4 |       256 |        0.006136 |   0.096746 |    15.767 |
|            8 |           4 |       512 |        0.006136 |   0.096876 |    15.788 |
|            8 |           4 |      1024 |        0.006084 |   0.096668 |    15.889 |
|            8 |           4 |      2048 |        0.006292 |   0.096694 |    15.368 |
|            8 |           8 |       256 |        0.007124 |   0.102128 |    14.336 |
|            8 |           8 |       512 |        0.00702  |   0.102336 |    14.578 |
|            8 |           8 |      1024 |        0.007124 |   0.10244  |    14.38  |
|            8 |           8 |      2048 |        0.007332 |   0.102388 |    13.965 |
|            8 |          16 |       256 |        0.006968 |   0.106236 |    15.246 |
|            8 |          16 |       512 |        0.00702  |   0.106236 |    15.133 |
|            8 |          16 |      1024 |        0.006916 |   0.106366 |    15.38  |
|            8 |          16 |      2048 |        0.00702  |   0.106158 |    15.122 |
|            8 |          32 |       256 |        0.00728  |   0.11258  |    15.464 |
|            8 |          32 |       512 |        0.007436 |   0.112372 |    15.112 |
|            8 |          32 |      1024 |        0.007332 |   0.11232  |    15.319 |
|            8 |          32 |      2048 |        0.007332 |   0.11232  |    15.319 |
|           16 |           4 |       256 |        0.00702  |   0.101556 |    14.467 |
|           16 |           4 |       512 |        0.007072 |   0.101816 |    14.397 |
|           16 |           4 |      1024 |        0.006916 |   0.101764 |    14.714 |
|           16 |           4 |      2048 |        0.007176 |   0.10166  |    14.167 |
|           16 |           8 |       256 |        0.007072 |   0.107432 |    15.191 |
|           16 |           8 |       512 |        0.007176 |   0.107588 |    14.993 |
|           16 |           8 |      1024 |        0.00702  |   0.107744 |    15.348 |
|           16 |           8 |      2048 |        0.007254 |   0.107432 |    14.81  |
|           16 |          16 |       256 |        0.007332 |   0.112424 |    15.333 |
|           16 |          16 |       512 |        0.007332 |   0.112502 |    15.344 |
|           16 |          16 |      1024 |        0.00728  |   0.11232  |    15.429 |
|           16 |          16 |      2048 |        0.00728  |   0.11232  |    15.429 |
|           16 |          32 |       256 |        0.007748 |   0.119912 |    15.477 |
|           16 |          32 |       512 |        0.0078   |   0.119652 |    15.34  |
|           16 |          32 |      1024 |        0.007696 |   0.120666 |    15.679 |
|           16 |          32 |      2048 |        0.007852 |   0.121264 |    15.444 |
|           32 |           4 |       256 |        0.007072 |   0.106496 |    15.059 |
|           32 |           4 |       512 |        0.007332 |   0.106236 |    14.489 |
|           32 |           4 |      1024 |        0.007072 |   0.106444 |    15.051 |
|           32 |           4 |      2048 |        0.007228 |   0.106236 |    14.698 |
|           32 |           8 |       256 |        0.007436 |   0.113152 |    15.217 |
|           32 |           8 |       512 |        0.007384 |   0.112944 |    15.296 |
|           32 |           8 |      1024 |        0.007332 |   0.112788 |    15.383 |
|           32 |           8 |      2048 |        0.007384 |   0.112814 |    15.278 |
|           32 |          16 |       256 |        0.0078   |   0.120588 |    15.46  |
|           32 |          16 |       512 |        0.007904 |   0.12064  |    15.263 |
|           32 |          16 |      1024 |        0.007748 |   0.121212 |    15.644 |
|           32 |          16 |      2048 |        0.007852 |   0.12194  |    15.53  |
|           32 |          32 |       256 |        0.008944 |   0.142272 |    15.907 |
|           32 |          32 |       512 |        0.008944 |   0.14196  |    15.872 |
|           32 |          32 |      1024 |        0.008944 |   0.141804 |    15.855 |
|           32 |          32 |      2048 |        0.008996 |   0.142896 |    15.884 |
|           64 |           4 |       256 |        0.00754  |   0.112788 |    14.959 |
|           64 |           4 |       512 |        0.007488 |   0.47346  |    63.229 |
|           64 |           4 |      1024 |        0.007436 |   0.112788 |    15.168 |
|           64 |           4 |      2048 |        0.007436 |   0.112918 |    15.185 |
|           64 |           8 |       256 |        0.007904 |   0.48932  |    61.908 |
|           64 |           8 |       512 |        0.008008 |   0.46436  |    57.987 |
|           64 |           8 |      1024 |        0.007852 |   0.121368 |    15.457 |
|           64 |           8 |      2048 |        0.008008 |   0.1222   |    15.26  |
|           64 |          16 |       256 |        0.008944 |   0.142012 |    15.878 |
|           64 |          16 |       512 |        0.009074 |   0.14196  |    15.645 |
|           64 |          16 |      1024 |        0.009048 |   0.243204 |    26.879 |
|           64 |          16 |      2048 |        0.008788 |   0.142376 |    16.201 |
|           64 |          32 |       256 |        0.013572 |   0.197288 |    14.536 |
|           64 |          32 |       512 |        0.013156 |   0.197444 |    15.008 |
|           64 |          32 |      1024 |        0.013312 |   0.197028 |    14.801 |
|           64 |          32 |      2048 |        0.013208 |   0.198952 |    15.063 |

### CODE REVIEW

File: tests/test_fused_q_indexer_rope_hadamard_quant.py [hunk 1]
```diff
@@ -1,20 +1,100 @@
...
 
 import pytest
 import torch
 from sgl_kernel import fused_q_indexer_rope_hadamard_quant
-from utils import get_device
+
+pytestmark = pytest.mark.skipif(
+    not torch.xpu.is_available(), reason="XPU not available"
+)
+_FP8_E4M3_MAX = 448.0
```
Comment (mingfeima) [inline tests/test_fused_q_indexer_rope_hadamard_quant.py]: do we have utils.py to hold all these constants?

File: tests/test_fused_q_indexer_rope_hadamard_quant.py [hunk 2]
```diff
@@ -1,20 +1,100 @@
...
+    for _ in range(iters):
+        start = torch.xpu.Event(enable_timing=True)
+        end = torch.xpu.Event(enable_timing=True)
+        start.record()
+        fn()
+        end.record()
+        torch.xpu.synchronize()
+        times.append(start.elapsed_time(end))
+    times.sort()
+    return times[len(times) // 2]
```
Comment (mingfeima) [inline tests/test_fused_q_indexer_rope_hadamard_quant.py]: remove bench code from UT.

File: tests/test_fused_q_indexer_rope_hadamard_quant.py [hunk 3]
```diff
@@ -1,20 +1,100 @@
...
+    positions: torch.Tensor,
+) -> None:
+    B, H, head_dim = q_input.shape
+    assert head_dim == _HEAD_DIM
+    assert freqs_cis.shape[-1] == _ROPE_DIM
+    assert q_fp8.shape == (B, H, _HEAD_DIM)
+    assert weight.shape == (B, H)
+    assert weights_out.shape == (B, H, 1)
+    if B == 0:
+        return
```
Comment (mingfeima) [inline tests/test_fused_q_indexer_rope_hadamard_quant.py]: can B be 0?

File: tests/test_fused_q_indexer_rope_hadamard_quant.py [hunk 4]
```diff
@@ -1,20 +1,100 @@
...
+
+    x = q_input.float()
+
+    freq = freqs_cis[positions.long()].unsqueeze(1)  # (B, 1, 64)
```
Comment (mingfeima) [inline tests/test_fused_q_indexer_rope_hadamard_quant.py]:
```suggestion
    freq = freqs_cis[positions].unsqueeze(1)  # (B, 1, 64)
```

the `.long()` is implicit. torch indexr only accepts int64. it will implicitly converts int32 to int64.

make python code pythonic.

What you need to keep in mind is the implicit conversion, on many occasions it is an overhead.

File: tests/test_fused_q_indexer_rope_hadamard_quant.py [hunk 5]
```diff
@@ -66,6 +145,64 @@ def test_fused_q_indexer_rope_hadamard_quant_runs(batch_size):
...
+            weights_out,
+            weight_scale,
+            torch.view_as_real(freqs_cis_xpu).flatten(-2),
+            positions_xpu,
+        )
+    )
+
+    assert (
+        t_sycl < t_ref
+    ), f"SYCL ({t_sycl:.3f} ms) not faster than torch ({t_ref:.3f} ms)"
```
Comment (mingfeima) [inline tests/test_fused_q_indexer_rope_hadamard_quant.py]: do not mix unit test with perf test.

File: src/sycl/FusedQIndexerRopeHadamardQuant.cpp [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    at::Tensor& weights_out,
+    double weight_scale,
+    const at::Tensor& rope_cache,
+    const at::Tensor& positions) {
+  CHECK_DEVICE(q_input);
+  CHECK_DEVICE(q_fp8);
+  CHECK_DEVICE(weight);
+  CHECK_DEVICE(weights_out);
+  CHECK_DEVICE(rope_cache);
+  CHECK_DEVICE(positions);
```
Comment (mingfeima) [inline src/sycl/FusedQIndexerRopeHadamardQuant.cpp]: use CHECK_INPUT

File: src/sycl/FusedQIndexerRopeHadamardQuant.cpp [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+  TORCH_CHECK(q_input.dim() == 3, "q_input must be [B, H, 128]");
```
Comment (mingfeima) [inline src/sycl/FusedQIndexerRopeHadamardQuant.cpp]: * check rank:

```CHECK_DIM(3, x);   // x.dim() == 3```


* check shape：

```CHECK_SHAPE(hc_fn, hc_mult, hc_mult * D);```

* check same shape

```CHECK_SAME_SHAPE(a, b);```

File: src/sycl/FusedQIndexerRopeHadamardQuant.cpp [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+  DISPATCH_FLOAT_TYPES(q_input.scalar_type(), "fused_q_indexer_rope_hadamard_quant", [&]() {
+    using sycl_in_t = typename SyclInputType<scalar_t>::type;
+    if (positions.scalar_type() == at::kInt) {
+      launch_fused_q_indexer_rope_hadamard_quant<sycl_in_t, int32_t>(
+          q_input, q_fp8, weight, weights_out, weight_scale, rope_cache, positions);
+    } else {
+      launch_fused_q_indexer_rope_hadamard_quant<sycl_in_t, int64_t>(
+          q_input, q_fp8, weight, weights_out, weight_scale, rope_cache, positions);
+    }
```
Comment (mingfeima) [inline src/sycl/FusedQIndexerRopeHadamardQuant.cpp]: first of all, ask yourself what it is the feasible dtype for positions.

if both int32 and int64, then

```
AT_DISPATCH_INDEX_TYPES
```

File: src/sycl/FusedQIndexerRopeHadamardQuant.cpp [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    {
+      const float a0 = data[0];
+      const float a1 = data[1];
+      const float a2 = data[2];
+      const float a3 = data[3];
+      data[0] = a0 + a2;
+      data[1] = a1 + a3;
+      data[2] = a0 - a2;
+      data[3] = a1 - a3;
+    }
```
Comment (mingfeima) [inline src/sycl/FusedQIndexerRopeHadamardQuant.cpp:97]:
```suggestion
const float x0 = data[0] + data[1];
const float x1 = data[0] - data[1];
const float x2 = data[2] + data[3];
const float x3 = data[2] - data[3];

data[0] = x0 + x2;
data[1] = x1 + x3;
data[2] = x0 - x2;
data[3] = x1 - x3;
```

File: src/sycl/FusedQIndexerRopeHadamardQuant.cpp [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+template <>
+struct SyclInputType<at::BFloat16> {
+  using type = sycl::ext::oneapi::bfloat16;
+};
+
+template <>
+struct SyclInputType<float> {
+  using type = float;
+};
```
Comment (mingfeima) [inline src/sycl/FusedQIndexerRopeHadamardQuant.cpp]: you don't need this

```
SYCL_DISPATCH_FLOATING_TYPES_AND2(
    at::ScalarType::BFloat16, at::ScalarType::Half,
```

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: tests/test_fused_q_indexer_rope_hadamard_quant.py (hunk 2 above)
Comment (cyxlily) [reply tests/test_fused_q_indexer_rope_hadamard_quant.py]: Thanks for the comments. Removed.

File: src/sycl/FusedQIndexerRopeHadamardQuant.cpp (hunk 7 above)
Comment (cyxlily) [reply src/sycl/FusedQIndexerRopeHadamardQuant.cpp]: Thanks for the comment. Replaced.

File: tests/test_fused_q_indexer_rope_hadamard_quant.py (hunk 1 above)
Comment (cyxlily) [reply tests/test_fused_q_indexer_rope_hadamard_quant.py]: Thanks for the comments. Moved constants to tests/utils.py.

File: tests/test_fused_q_indexer_rope_hadamard_quant.py (hunk 3 above)
Comment (cyxlily) [reply tests/test_fused_q_indexer_rope_hadamard_quant.py]: In practice B is rarely 0 in DeepSeek V4's real forward path, there is a check for it: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/models/deepseek_v4.py#L1544

**Review by mingfeima — APPROVED** [review]
